### PR TITLE
sql: refactor fk backreference removal

### PR DESCRIPTION
**Pull request by postamar**
Previously, from the point of view of the outbound and inbound table
descriptors, the removal of a foreign key would be performed separately.
Ideally we'd want the OutboundFKs slice of one table descriptor and the
InboundFKs slice of the other to always be consistent, both in storage
(which is already the case) as well as in memory. This refactor commit
addresses this by making these changes closer together for the mutable
table descriptors.

Release note: None

### CODE REVIEW

**Comment by postamar** [comment]
I came across this while working on https://github.com/cockroachdb/cockroach/pull/60552 . Personally I found all the `OutboundFKs` and `InboundFKs` slice surgeries hard to follow, particularly those with `sliceIdx`. Logically we're removing the ref from both `OutboundFKs` and `InboundFKs` but it's not implemented that way for some reason (premature optimization?) 
This is just my opinion but I'm curious to hear whether this commit I'm proposing is a genuine improvement worth the risk or instead rather gratuitous and forgettable.

**Review by ajwerner — COMMENTED** [review]
was this ready to go? I think it slipped my radar. May need to wait until the branch cut now. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60597#-:-MUBvIDp3h6YdVS87lKN:b-mz877s)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by postamar** [comment]
Yes, it should be ready to go. It's not time-sensitive though, it can wait.

**Review by ajwerner — COMMENTED** [review]
I'm not so sure I'm keen on this pattern. Can we abstract it better either with a copy or a helper somewhere? It is not obvious without a comment that this is all motivated by the fact that we're mutating the slice as we iterate it. 

> Reviewed 1 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60597#-:-MVX70CH0-owicZ53QBh:brz2xx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

**Comment by postamar** [comment]
I think you're right, there must be a better way.
